File: src/android/input/StartCallInput.java
package com.bandyer.cordova.plugin.input;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

import com.bandyer.cordova.plugin.Constants;

import static com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO;
import static com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO_UPGRADABLE;
import static com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO_VIDEO;

public class StartCallInput {

    private final ArrayList<String> mCalleeList;
    private String mJoinUrl;
    private boolean isRecordingEnabled;
    private CallType callType;

    public static StartCallInput createFrom(JSONArray argsArray) throws PluginInputNotValidException {
        StartCallInput startInput = new StartCallInput();
        try {
            JSONObject args = (JSONObject) argsArray.get(0);
            JSONArray callee = args.has(Constants.ARG_CALLEE) ? args.getJSONArray(Constants.ARG_CALLEE) : new JSONArray();
            String joinUrl = args.has(Constants.ARG_JOIN_URL) ? args.getString(Constants.ARG_JOIN_URL) : "";
            boolean isThereACallee = callee != null && callee.length() > 0;
            boolean isThereAJoinUrl = joinUrl != null && !joinUrl.equals("");
            if(!isThereACallee && !isThereAJoinUrl) {
                throw new PluginInputNotValidException(Constants.ARG_CALLEE + "and " + Constants.ARG_JOIN_URL + " cannot be null");
            }
            if(isThereACallee){
                int len = callee.length();
                for (int i=0;i<len;i++){
                    startInput.addCallee(callee.getString(i));
                }
            }
            if(isThereAJoinUrl){
                startInput.setJoinUrl(joinUrl);
            }
            boolean rec = args.has(Constants.ARG_RECORDING) ? args.getBoolean(Constants.ARG_RECORDING) : false;
            startInput.setRecordingEnabled(rec);
            String type = args.has(Constants.ARG_CALL_TYPE) ? args.getString(Constants.ARG_CALL_TYPE) : VALUE_CALL_TYPE_AUDIO_VIDEO;
            if(VALUE_CALL_TYPE_AUDIO.equals(type)){
                startInput.setCallType(CallType.AUDIO);
            } else if(VALUE_CALL_TYPE_AUDIO_UPGRADABLE.equals(type)){
                startInput.setCallType(CallType.AUDIO_UPGRADABLE);
            } else {
                startInput.setCallType(CallType.AUDIO_VIDEO);
            }
            return startInput;
        }catch (Throwable t) {
            throw new PluginInputNotValidException("error on InitInput " + t.getMessage(), t);
        }
    }

    private StartCallInput(){
        mCalleeList = new ArrayList<>();
    }

    public ArrayList<String> getCalleeList() {
        return mCalleeList;
    }

    public void addCallee(String callee) {
        mCalleeList.add(callee);
    }

    public String getJoinUrl() {
        return mJoinUrl;
    }

    public void setJoinUrl(String joinUrl) {
        mJoinUrl = joinUrl;
    }

    public boolean hasJoinUrl() {
        return mJoinUrl != null && mJoinUrl.length() > 0;
    }

    public boolean isRecordingEnabled() {
        return isRecordingEnabled;
    }

    public void setRecordingEnabled(boolean recordingEnabled) {
        isRecordingEnabled = recordingEnabled;
    }

    public CallType getCallType() {
        return callType;
    }

    public void setCallType(CallType callType) {
        this.callType = callType;
    }
}

File: src/android/cordova_plugin/src/main/java/com/bandyer/cordova/plugin/input/StartCallInput.kt
package com.bandyer.cordova.plugin.input

import org.json.JSONArray
import org.json.JSONObject

import java.util.ArrayList

import com.bandyer.cordova.plugin.Constants

import com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO
import com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO_UPGRADABLE
import com.bandyer.cordova.plugin.Constants.VALUE_CALL_TYPE_AUDIO_VIDEO

class StartCallInput private constructor() {

    val calleeList: ArrayList<String>
    var joinUrl: String? = null
    var isRecordingEnabled: Boolean = false
    var callType: CallType? = null

    init {
        calleeList = ArrayList()
    }

    fun addCallee(callee: String) {
        calleeList.add(callee)
    }

    fun hasJoinUrl(): Boolean {
        return joinUrl != null && joinUrl!!.length > 0
    }

    companion object {

        @Throws(PluginInputNotValidException::class)
        fun createFrom(argsArray: JSONArray): StartCallInput {
            val startInput = StartCallInput()
            try {
                val args = argsArray.get(0) as JSONObject
                val callee = if (args.has(Constants.ARG_CALLEE)) args.getJSONArray(Constants.ARG_CALLEE) else JSONArray()
                val joinUrl = if (args.has(Constants.ARG_JOIN_URL)) args.getString(Constants.ARG_JOIN_URL) else ""
                val isThereACallee = callee != null && callee.length() > 0
                val isThereAJoinUrl = joinUrl != null && joinUrl != ""
                if (!isThereACallee && !isThereAJoinUrl) {
                    throw PluginInputNotValidException(Constants.ARG_CALLEE + "and " + Constants.ARG_JOIN_URL + " cannot be null")
                }
                if (isThereACallee) {
                    val len = callee.length()
                    for (i in 0 until len) {
                        startInput.addCallee(callee.getString(i))
                    }
                }
                if (isThereAJoinUrl) {
                    startInput.joinUrl = joinUrl
                }
                val rec = if (args.has(Constants.ARG_RECORDING)) args.getBoolean(Constants.ARG_RECORDING) else false
                startInput.isRecordingEnabled = rec
                val type = if (args.has(Constants.ARG_CALL_TYPE)) args.getString(Constants.ARG_CALL_TYPE) else VALUE_CALL_TYPE_AUDIO_VIDEO
                if (VALUE_CALL_TYPE_AUDIO == type) {
                    startInput.callType = CallType.AUDIO
                } else if (VALUE_CALL_TYPE_AUDIO_UPGRADABLE == type) {
                    startInput.callType = CallType.AUDIO_UPGRADABLE
                } else {
                    startInput.callType = CallType.AUDIO_VIDEO
                }
                return startInput
            } catch (t: Throwable) {
                throw PluginInputNotValidException("error on InitInput " + t.message, t)
            }

        }
    }
}
